core/event.cpp

```diff
@@ -93,5 +96,133 @@ Variable broadcast_weights(const DataArrayConstView &events) {
+      copy_if_args<int64_t, int32_t>, copy_if_args<int32_t, int32_t>,
+      copy_if_args<double, int64_t>, copy_if_args<float, int64_t>,
+      copy_if_args<int64_t, int64_t>, copy_if_args<int32_t, int64_t>>>(
+      var, select,
+      overloaded{
+          transform_flags::expect_no_variance_arg<1>,
+          [](const auto &var_, const auto &select_) {
+            using VarT = std::decay_t<decltype(var_)>;
+            using Events = event_list<typename VarT::value_type>;
+            const auto size = scipp::size(select_);
+            if constexpr (detail::is_ValuesAndVariances_v<VarT>) {
+              std::pair<Events, Events> out;
+              out.first.reserve(size);
+              out.second.reserve(size);
+              for (const auto i : select_) {
+                out.first.push_back(var_.values[i]);
+                out.second.push_back(var_.variances[i]);
+              }
+              return out;
+            } else {
+              Events out;
+              out.reserve(size);
+              for (const auto i : select_)
+                out.push_back(var_[i]);
+              return out;
+            }
+          },
+          [](const units::Unit &var_, const units::Unit &) { return var_; }});
+};
+
+template <class T>
+const auto make_select = [](const DataArrayConstView &array, const Dim dim,
+                            const VariableConstView &interval) {
+  return transform<
+      std::tuple<make_select_args<double>, make_select_args<float>,
+                 make_select_args<int64_t>, make_select_args<int32_t>>>(
+      array.coords()[dim], subspan_view(interval, dim),
+      overloaded{transform_flags::expect_no_variance_arg<0>,
+                 transform_flags::expect_no_variance_arg<1>,
+                 [](const auto &coord_, const auto &interval_) {
+                   const auto low = interval_[0];
+                   const auto high = interval_[1];
+                   const auto size = scipp::size(coord_);
+                   event_list<T> select_;
+                   for (scipp::index i = 0; i < size; ++i)
+                     if (coord_[i] >= low && coord_[i] < high)
+                       select_.push_back(i);
+                   return select_;
+                 },
+                 [](const units::Unit &coord_, const units::Unit &interval_) {
+                   expect::equals(coord_, interval_);
+                   return units::Unit(units::dimensionless);
+                 }});
+};
+
+} // namespace filter_detail
+
+DataArray filter(const DataArrayConstView &array, const Dim dim,
+                 const VariableConstView &interval,
+                 const AttrPolicy attrPolicy) {
+  using namespace filter_detail;
+  const auto &max_event_list_length = max(sizes(array.coords()[dim]));
+  const bool need_64bit_indices =
+      max_event_list_length.values<scipp::index>()[0] >
+      std::numeric_limits<int32_t>::max();
+  const auto select = need_64bit_indices
```

Comment:
Your strategy is to separate the selection from the application. This would seem particularly useful if the selection was to be reused. In current filter it doesn't seem to be.

Reply:
It is reused 2 or 3 times typically: For the coord, such as pulse-time, for the TOF coord, and for the data.

Reply:
This is also needed to reduce a combinatoric explosion of type combinations, for edges, data, multiple coords, ...

Reply:
Good point.